This is the dataset from Kaggle Dataset 

In [1]:
import pandas as pd

establishment_df=pd.read_csv("data/Grab SG Restaurants.csv")

establishment_df.head()


id_source             name                     address    country  \
0         SGDD01085    Texas Chicken    Texas Chicken - VivoCity  Singapore   
1  4-CYTDLPUJEP53N6           Subway           Subway - Vivocity  Singapore   
2         SGDD05605         LiHO Tea        LiHO Tea - Vivo City  Singapore   
3         SGDD07548         Wingstop         Wingstop - VivoCity  Singapore   
4  4-CY2ZLZLTT2EDJ2  Le Shrimp Ramen  Le Shrimp Ramen - Vivocity  Singapore   

                                             cuisine currency  delivery_cost  \
0  ["Western", "Fast Food", "Halal", "Chicken", "...      SGD          570.0   
1  ["Healthy", "Sandwiches", "Fast Food", "Halal"...      SGD          570.0   
2  ["Same Prices In-Store", "Beverages", "Local",...      SGD          560.0   
3  ["Halal", "Chicken", "Fast Food", "Burger", "W...      SGD          560.0   
4          ["Japanese", "Ramen", "Noodles", "Asian"]      SGD          570.0   

        lat         lon                                      opening_hours  \
0  1.264792  103.822103  {"open": true, "displayedHours": "10:30-21:30"...   
1  1.264741  103.822072  {"open": true, "displayedHours": "12:00-19:00"...   
2  1.264309  103.821369  {"open": true, "displayedHours": "11:30-20:45"...   
3  1.263527  103.821712  {"open": true, "displayedHours": "11:00-21:00"...   
4  1.264810  103.822554  {"open": true, "displayedHours": "10:30-21:00"...   

                                           image_url  radius  rating  \
0  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     3.8   
1  https://d1sag4ddilekf6.cloudfront.net/compress...    3000     4.2   
2  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     4.5   
3  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     4.4   
4  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     4.6   

   reviews_nr   delivery_options  \
0      1591.0      ONLY_DELIVERY   
1       197.0      ONLY_DELIVERY   
2       353.0  DELIVERY_TAKEAWAY   
3      1391.0      ONLY_DELIVERY   
4       309.0  DELIVERY_TAKEAWAY   

                                               promo loc_type delivery_by  \
0  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
1  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
2  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
3  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
4  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   

   delivery_time  
0           43.0  
1           42.0  
2           41.0  
3           45.0  
4           44.0

## Exploratory Data Analysis

Simple exploratory Data Analysis to understand the data

In [2]:
establishment_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16136 entries, 0 to 16135
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id_source         16136 non-null  object 
 1   name              16130 non-null  object 
 2   address           16136 non-null  object 
 3   country           16136 non-null  object 
 4   cuisine           16134 non-null  object 
 5   currency          16117 non-null  object 
 6   delivery_cost     16085 non-null  float64
 7   lat               16136 non-null  float64
 8   lon               16136 non-null  float64
 9   opening_hours     16136 non-null  object 
 10  image_url         16125 non-null  object 
 11  radius            16136 non-null  int64  
 12  rating            14004 non-null  float64
 13  reviews_nr        13602 non-null  float64
 14  delivery_options  16136 non-null  object 
 15  promo             8777 non-null   object 
 16  loc_type          16136 non-null  object

So the dataset of 16136 rows of grab restaurants
you can see from the count that there are empty values for each column

In [3]:
establishment_df.describe()

delivery_cost           lat           lon        radius        rating  \
count   16085.000000  16136.000000  16136.000000  16136.000000  14004.000000   
mean      984.926640      1.338491    103.843448   6871.493183      4.238346   
std       656.544752      0.045269      0.062123   6170.013565      0.619217   
min       199.000000      1.230454    103.625677    578.000000     -1.000000   
25%       450.000000      1.304839    103.806196   4999.000000      4.000000   
50%       680.000000      1.332154    103.846004   5000.000000      4.300000   
75%      1610.000000      1.362576    103.885674   5000.000000      4.600000   
max     14000.000000      1.469011    103.997253  25000.000000      5.000000   

         reviews_nr  delivery_time  
count  13602.000000   15953.000000  
mean     343.591604      37.552811  
std      846.893054      15.728705  
min        1.000000      15.000000  
25%       16.000000      28.000000  
50%       74.000000      34.000000  
75%      303.000000      42.000000  
max    21113.000000     169.000000

# Data Cleaning on 'address' column: SAILA
### Which involves data inspection, Data correction, Data standardization and Data transformation.

- Remove duplicates-> the address column values are keyed in together with the establishment name seperated by "-"
- Standardize the wording -> currently the values entries are inconsistent cases: 
    example : 'Vivocity' , 'VivoCity'
- Inconsistent entry -> for the same address it was entered inconsistantly
    example: 'Vivocity', 'Vivo City' , 'Vivocity Shopping mall'
- Need to filter and transform the address entries to avoid unnecessary disticnt values.

In [4]:
#Splitting the data .Removing values before '-'
establishment_df['address'] = establishment_df['address'].apply(lambda x: x.split('-')[-1].strip())

# Display the modified DataFrame
print(establishment_df)


              id_source                         name  \
0             SGDD01085                Texas Chicken   
1      4-CYTDLPUJEP53N6                       Subway   
2             SGDD05605                     LiHO Tea   
3             SGDD07548                     Wingstop   
4      4-CY2ZLZLTT2EDJ2              Le Shrimp Ramen   
...                 ...                          ...   
16131  4-CZNFTYKKVA3UN2                 Minjiangkueh   
16132         SGDD01192  The Modern Izakaya by Tatsu   
16133         SGDD00961               Raj Restaurant   
16134  4-C2AHJVATN7C2RE   Albismi Indian Muslim Food   
16135  4-C2DXFEK2R6KJTJ  Zam Zam Indian Special Food   

                        address    country  \
0                      VivoCity  Singapore   
1                      Vivocity  Singapore   
2                     Vivo City  Singapore   
3                      VivoCity  Singapore   
4                      Vivocity  Singapore   
...                         ...        ...   
16131

In [5]:
establishment_df.head()

id_source             name    address    country  \
0         SGDD01085    Texas Chicken   VivoCity  Singapore   
1  4-CYTDLPUJEP53N6           Subway   Vivocity  Singapore   
2         SGDD05605         LiHO Tea  Vivo City  Singapore   
3         SGDD07548         Wingstop   VivoCity  Singapore   
4  4-CY2ZLZLTT2EDJ2  Le Shrimp Ramen   Vivocity  Singapore   

                                             cuisine currency  delivery_cost  \
0  ["Western", "Fast Food", "Halal", "Chicken", "...      SGD          570.0   
1  ["Healthy", "Sandwiches", "Fast Food", "Halal"...      SGD          570.0   
2  ["Same Prices In-Store", "Beverages", "Local",...      SGD          560.0   
3  ["Halal", "Chicken", "Fast Food", "Burger", "W...      SGD          560.0   
4          ["Japanese", "Ramen", "Noodles", "Asian"]      SGD          570.0   

        lat         lon                                      opening_hours  \
0  1.264792  103.822103  {"open": true, "displayedHours": "10:30-21:30"...   
1  1.264741  103.822072  {"open": true, "displayedHours": "12:00-19:00"...   
2  1.264309  103.821369  {"open": true, "displayedHours": "11:30-20:45"...   
3  1.263527  103.821712  {"open": true, "displayedHours": "11:00-21:00"...   
4  1.264810  103.822554  {"open": true, "displayedHours": "10:30-21:00"...   

                                           image_url  radius  rating  \
0  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     3.8   
1  https://d1sag4ddilekf6.cloudfront.net/compress...    3000     4.2   
2  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     4.5   
3  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     4.4   
4  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     4.6   

   reviews_nr   delivery_options  \
0      1591.0      ONLY_DELIVERY   
1       197.0      ONLY_DELIVERY   
2       353.0  DELIVERY_TAKEAWAY   
3      1391.0      ONLY_DELIVERY   
4       309.0  DELIVERY_TAKEAWAY   

                                               promo loc_type delivery_by  \
0  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
1  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
2  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
3  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
4  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   

   delivery_time  
0           43.0  
1           42.0  
2           41.0  
3           45.0  
4           44.0

#### To Understand the dataframe entries-Distinct values at address column

In [6]:

distinct_addresses = establishment_df['address'].unique()
address_counts = establishment_df['address'].value_counts()

# Display distinct values and their counts
print("Distinct Addresses:")
print(distinct_addresses)
print("\nCount of Distinct Addresses:")
print(address_counts.head(50))


Distinct Addresses:
['VivoCity' 'Vivocity' 'Vivo City' ... 'Wilmer Place' '18)'
 'SOHO @ Farrer']

Count of Distinct Addresses:
address
Northpoint City                 90
Jurong Point                    86
Waterway Point                  81
Causeway Point                  76
Plaza Singapura                 74
Changi City Point               74
Westgate                        70
Geylang Road                    70
Suntec City Mall                65
East Coast Road                 62
Tiong Bahru Plaza               60
Great World City                59
Compass One                     59
Vivocity                        58
Hillion Mall                    57
NEX                             57
Our Tampines Hub                55
JEM                             55
Bedok Mall                      53
Tampines Mall                   49
New Upper Changi Road           48
Century Square                  45
Jewel Changi Airport            45
Harbourfront Centre             45
Vivo City               

#### Total Distinct values count

In [7]:
# Assuming 'establishment_df' is your DataFrame containing the data
total_distinct_addresses = establishment_df['address'].nunique()

# Display the total number of distinct addresses
print("Total distinct addresses:", total_distinct_addresses)


Total distinct addresses: 4998


### Standardize all the add with tittle case.

In [8]:
# Assuming 'establishment_df' is your DataFrame containing the data
establishment_df['address'] = establishment_df['address'].apply(lambda x: x.title())

# Display the modified DataFrame
establishment_df.head()


id_source             name    address    country  \
0         SGDD01085    Texas Chicken   Vivocity  Singapore   
1  4-CYTDLPUJEP53N6           Subway   Vivocity  Singapore   
2         SGDD05605         LiHO Tea  Vivo City  Singapore   
3         SGDD07548         Wingstop   Vivocity  Singapore   
4  4-CY2ZLZLTT2EDJ2  Le Shrimp Ramen   Vivocity  Singapore   

                                             cuisine currency  delivery_cost  \
0  ["Western", "Fast Food", "Halal", "Chicken", "...      SGD          570.0   
1  ["Healthy", "Sandwiches", "Fast Food", "Halal"...      SGD          570.0   
2  ["Same Prices In-Store", "Beverages", "Local",...      SGD          560.0   
3  ["Halal", "Chicken", "Fast Food", "Burger", "W...      SGD          560.0   
4          ["Japanese", "Ramen", "Noodles", "Asian"]      SGD          570.0   

        lat         lon                                      opening_hours  \
0  1.264792  103.822103  {"open": true, "displayedHours": "10:30-21:30"...   
1  1.264741  103.822072  {"open": true, "displayedHours": "12:00-19:00"...   
2  1.264309  103.821369  {"open": true, "displayedHours": "11:30-20:45"...   
3  1.263527  103.821712  {"open": true, "displayedHours": "11:00-21:00"...   
4  1.264810  103.822554  {"open": true, "displayedHours": "10:30-21:00"...   

                                           image_url  radius  rating  \
0  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     3.8   
1  https://d1sag4ddilekf6.cloudfront.net/compress...    3000     4.2   
2  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     4.5   
3  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     4.4   
4  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     4.6   

   reviews_nr   delivery_options  \
0      1591.0      ONLY_DELIVERY   
1       197.0      ONLY_DELIVERY   
2       353.0  DELIVERY_TAKEAWAY   
3      1391.0      ONLY_DELIVERY   
4       309.0  DELIVERY_TAKEAWAY   

                                               promo loc_type delivery_by  \
0  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
1  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
2  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
3  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   
4  Use 'TEATIME' for FREE delivery between 3PM to...     FOOD        GRAB   

   delivery_time  
0           43.0  
1           42.0  
2           41.0  
3           45.0  
4           44.0

#### Cleaning the [address] column further by changing the case to tittle form.The distinct count reduced from 4998 to 4883

In [9]:
#To find the total distinct value after changing the values in address column to tittle case.
total_distinct_addresses = establishment_df['address'].nunique()

# Display the total number of distinct addresses
print("Total distinct addresses:", total_distinct_addresses)

Total distinct addresses: 4883


### Filtering out the distinct data using alphabets to verify the values in the column and to do necessary transformation needed if the data was not input in accurately.

In [10]:
# Assuming 'establishment_df' is your DataFrame containing the data

# Step 1: Filter the DataFrame
filtered_df = establishment_df[establishment_df['address'].str.lower().str.startswith(('a'))]

# Step 2: Extract distinct values from the filtered DataFrame
distinct_addresses_filtered = filtered_df['address'].unique()

# Step 3: Sort the distinct values alphabetically
distinct_addresses_filtered_sorted = sorted(distinct_addresses_filtered)

# Display the sorted distinct values
print("Distinct addresses starting with 'a':")
for address in distinct_addresses_filtered_sorted:
    print(address)


Distinct addresses starting with 'a':
A' Posh Bizhub
A'Posh Bizhub
Abc Brickworks Food Centre
Abc Brickworks Market & Food Centre
Abc Brickworks Market & Food Centre [Stall 113]
Abc Brickworks Market & Food Centre [Stall 120]
Abc Brickworks Market & Food Centre [Stall 12]
Abc Brickworks Market & Food Centre [Stall 143]
Abc Brickworks Market & Food Centre [Stall 15]
Abc Brickworks Market & Food Centre [Stall 18]
Abc Brickworks Market And Food Centre [Stall 10]
Abc Brickworks Market And Food Centre [Stall 2]
Abc Brickworks Market And Food Centre [Stall 8]
Acacia Breeze @ Yishun
Acacia Court
Academia
Academy Of Singapore Teachers
Ace Community Club
Adam Road Food Centre
Adda
Adelphi Park Estate
Admiralty
Admiralty Link
Admiralty Street
Admiralty [Islandwide Delivery]
Admiralty[Islandwide Delivery]
Admirax
Admirax [Islandwide Delivery]
Agape Village
Ah Hood
Ah Hood Road
Ah Wang Kway Chap
Air View Building
Airport Boulevard
Airport Boulevard [Islandwide Delivery]
Airview Building
Albert Cen

### Tried using difflib library to standardize the entries up to 80% matching.( Didn't work)

In [11]:
import difflib

# Filter the DataFrame to include only addresses starting with 'a'
addresses_starting_with_a = establishment_df[establishment_df['address'].str.lower().str.startswith('a')]['address'].unique()

# Define a function to find the best match for each address starting with 'a'
def standardize_address(address):
    matches = difflib.get_close_matches(address, addresses_starting_with_a, n=1, cutoff=0.7)
    if matches:
        return matches[0]
    else:
        return address

# Apply the standardization function to addresses starting with 'a'
standardized_addresses_starting_with_a = [standardize_address(address) for address in addresses_starting_with_a]

# Display the original and standardized addresses
print("Original Addresses:")
print(addresses_starting_with_a)

print("\nStandardized Addresses:")
print(standardized_addresses_starting_with_a)


Original Addresses:
['Anson House' 'Alkaff Mansion' 'Academia' 'Alexandra Road' 'Altez'
 'Anson' 'Apex @ Henderson'
 'Ayer Rajah Food Centre [Stall 10] [Islandwide Delivery]'
 'Alexandra Retail Centre' 'Alexandra Retail Centre [Islandwide Delivery]'
 'Asia Square Tower 2 [Islandwide Delivery]' 'Arc [Islandwide Delivery]'
 'Amoy Street Food Centre [Islandwide Delivery]'
 'Ayer Rajah Food Centre [Islandwide Delivery]' 'Ayer Rajah Crescent'
 'Alexandra Village Food Centre' 'Alexandra Central Mall'
 'Alexandra Hawker Centre [Islandwide Delivery]'
 'Anchorpoint Shopping Centre' 'Anchor Point' 'Alexandra Central'
 'Avocado & Natural Foods  (Anchorpoint Shopping Centre)' 'Arc'
 'Ayer Rajah Food Centre 1' 'Alexandra Retail Centre Psa Building'
 'Abc Brickworks Market & Food Centre' 'Ascent' 'Alexandra' 'Anchorpoint'
 'Abc Brickworks Market & Food Centre [Stall 12]'
 'Abc Brickworks Market & Food Centre [Stall 15]'
 'Abc Brickworks Market And Food Centre [Stall 8]'
 'Ayer Rajah Food Centre I' '

### The result shows the total number of entries are the same before and after 'standardization' using the 'difflib' library

In [12]:
# Count the total number of entries before standardization
total_entries_before = len(addresses_starting_with_a)

# Count the total number of entries after standardization
total_entries_after = len(set(standardized_addresses_starting_with_a))

# Print the total number of entries before and after standardization
print("Total entries with addresses starting with 'a' before standardization:", total_entries_before)
print("Total entries with addresses starting with 'a' after standardization:", total_entries_after)

Total entries with addresses starting with 'a' before standardization: 237
Total entries with addresses starting with 'a' after standardization: 237


### Standardize all the entries that have '[xxx]' and replace all the 'And' with '&' for standardization

In [13]:
import re

# Define a function to standardize addresses and count changes
def standardize_address_and_count(address):
    # Initialize counter for changes
    num_changes = 0
    
    # Remove '[Stall xx]' part using regular expression
    standardized_address = re.sub(r'\[.*?\]', '', address)
    
    # Replace 'And' with '&' and remove leading/trailing spaces
    standardized_address = standardized_address.replace('And', '&').strip()
    
    # Replace any occurrence of 'Abc Brickworks' with 'Abc Brickworks Market & Food Centre'
    if 'Abc Brickworks' in standardized_address:
        standardized_address = standardized_address.replace('Abc Brickworks', 'Abc Brickworks Market & Food Centre')
        num_changes += 1
    
    return standardized_address, num_changes

# Apply the standardization function to the 'address' column of the DataFrame
establishment_df['standardized_address'], num_changes = zip(*establishment_df['address'].apply(standardize_address_and_count))

# Print the total count of entries that were changed
print("Total count of entries changed:", sum(num_changes))

# Print the list of entries that were changed
changed_entries = establishment_df[establishment_df['standardized_address'] != establishment_df['address']]
print("\nList of entries that were changed:")
for index, row in changed_entries.iterrows():
    print(row['address'], '->', row['standardized_address'])


Total count of entries changed: 11

List of entries that were changed:
Heytea Vivo City [Islandwide Delivery] -> Heytea Vivo City
Vivocity [Islandwide Delivery] -> Vivocity
Vivocity [Islandwide Delivery] -> Vivocity
Harbourfront Centre [Islandwide Delivery] -> Harbourfront Centre
Resorts World Sentosa [Islandwide Delivery] -> Resorts World Sentosa
W Singapore Sentosa Cove [Islandwide Delivery] -> W Singapore Sentosa Cove
Telok Blangah Rise [Islandwide Delivery] -> Telok Blangah Rise
33 Kampong Bahru Road [Islandwide Delivery] -> 33 Kampong Bahru Road
Lim Teck Kim Road [Islandwide Delivery] -> Lim Teck Kim Road
Tanjong Pagar Complex [Islandwide Delivery] -> Tanjong Pagar Complex
Neil Road [Islandwide Delivery] -> Neil Road
Everton Park[Islandwide Delivery] -> Everton Park
Neil Road [Islandwide Delivery] -> Neil Road
Tras Link [Islandwide Delivery] -> Tras Link
Telok Blangah Drive [Islandwide Delivery] -> Telok Blangah Drive
Yong Siak View [Islandwide Delivery] -> Yong Siak View
Yeshangh

In [14]:
#To find the total distinct value after changing the values in address column to tittle case.
total_distinct_addresses = establishment_df['address'].nunique()

# Display the total number of distinct addresses
print("Total distinct addresses:", total_distinct_addresses)

Total distinct addresses: 4883


In [15]:
# Step 1: Filter the DataFrame
filtered_df = establishment_df[establishment_df['address'].str.lower().str.startswith(('a'))]

# Step 2: Extract distinct values from the filtered DataFrame
distinct_addresses_filtered = filtered_df['address'].unique()

# Step 3: Sort the distinct values alphabetically
distinct_addresses_filtered_sorted = sorted(distinct_addresses_filtered)

# Display the sorted distinct values
print("Distinct addresses starting with 'a':")
for address in distinct_addresses_filtered_sorted:
    print(address)

Distinct addresses starting with 'a':
A' Posh Bizhub
A'Posh Bizhub
Abc Brickworks Food Centre
Abc Brickworks Market & Food Centre
Abc Brickworks Market & Food Centre [Stall 113]
Abc Brickworks Market & Food Centre [Stall 120]
Abc Brickworks Market & Food Centre [Stall 12]
Abc Brickworks Market & Food Centre [Stall 143]
Abc Brickworks Market & Food Centre [Stall 15]
Abc Brickworks Market & Food Centre [Stall 18]
Abc Brickworks Market And Food Centre [Stall 10]
Abc Brickworks Market And Food Centre [Stall 2]
Abc Brickworks Market And Food Centre [Stall 8]
Acacia Breeze @ Yishun
Acacia Court
Academia
Academy Of Singapore Teachers
Ace Community Club
Adam Road Food Centre
Adda
Adelphi Park Estate
Admiralty
Admiralty Link
Admiralty Street
Admiralty [Islandwide Delivery]
Admiralty[Islandwide Delivery]
Admirax
Admirax [Islandwide Delivery]
Agape Village
Ah Hood
Ah Hood Road
Ah Wang Kway Chap
Air View Building
Airport Boulevard
Airport Boulevard [Islandwide Delivery]
Airview Building
Albert Cen

#### 2nd attempt on data standardization (it created a seperate columns instead)

In [16]:
import re

# Define a function to standardize addresses and count changes
def standardize_address_and_count(address):
    # Initialize counter for changes
    num_changes = 0
    
    # Remove '[Stall xx]' part using regular expression
    standardized_address = re.sub(r'\[.*?\]', '', address)
    
    # Replace 'And' with '&' and remove leading/trailing spaces
    standardized_address = standardized_address.replace('And', '&').strip()
    
    # Replace any occurrence of 'Abc Brickworks' with 'Abc Brickworks Market & Food Centre'
    if 'Abc Brickworks' in standardized_address:
        standardized_address = standardized_address.replace('Abc Brickworks', 'Abc Brickworks Market & Food Centre')
        num_changes += 1
    
    return standardized_address, num_changes

# Apply the standardization function to the 'address' column of the DataFrame
establishment_df['address'], num_changes = zip(*establishment_df['address'].apply(standardize_address_and_count))

# Print the total count of entries that were changed
print("Total count of entries changed:", sum(num_changes))

# Print the list of entries that were changed
changed_entries = establishment_df[establishment_df['standardized_address'] != establishment_df['address']]

print("\nList of entries that were changed:")
for index, row in changed_entries.iterrows():
    print(row['address'], '->', row['standardized_address'])


Total count of entries changed: 11

List of entries that were changed:


In [17]:
# Step 1: Filter the DataFrame
filtered_df = establishment_df[establishment_df['address'].str.lower().str.startswith(('a'))]

# Step 2: Extract distinct values from the filtered DataFrame
distinct_addresses_filtered = filtered_df['address'].unique()

# Step 3: Sort the distinct values alphabetically
distinct_addresses_filtered_sorted = sorted(distinct_addresses_filtered)

# Display the sorted distinct values
print("Distinct addresses starting with 'a':")
for address in distinct_addresses_filtered_sorted:
    print(address)

Distinct addresses starting with 'a':
A' Posh Bizhub
A'Posh Bizhub
Abc Brickworks Market & Food Centre Food Centre
Abc Brickworks Market & Food Centre Market & Food Centre
Acacia Breeze @ Yishun
Acacia Court
Academia
Academy Of Singapore Teachers
Ace Community Club
Adam Road Food Centre
Adda
Adelphi Park Estate
Admiralty
Admiralty Link
Admiralty Street
Admirax
Agape Village
Ah Hood
Ah Hood Road
Ah Wang Kway Chap
Air View Building
Airport Boulevard
Airview Building
Albert Centre
Alexandra
Alexandra Central
Alexandra Central Mall
Alexandra Hawker Centre
Alexandra Hospital
Alexandra Retail Centre
Alexandra Retail Centre Psa Building
Alexandra Road
Alexandra Technopark
Alexandra Village Food Centre
Aljunied
Aljunied 119 Food House
Aljunied Ave 2
Aljunied Ave 3
Aljunied Ave 4
Aljunied Ave 5
Aljunied Avenue 2
Aljunied B
Aljunied Crescent
Aljunied Industrial Complex
Aljunied Industrial Estate
Aljunied Mrt
Aljunied Mrt Station
Aljunied Road
Alkaff Crescent
Alkaff Mansion
Alkaff Vista
Alocassia

### The distinct Values reduced from 4883 to 4029

In [18]:
#To find the total distinct value after standardizing the data with removal of '[xxx]' and change 'And' to '&'
total_distinct_addresses = establishment_df['address'].nunique()

# Display the total number of distinct addresses
print("Total distinct addresses:", total_distinct_addresses)

Total distinct addresses: 4029


### Further standiardize the entries in address column by removing whatever additional information after the add like brackets and such, replace all the 'Amk' to 'Ang Mo Kio'

In [19]:
import re

# Define a function to standardize addresses
def standardize_address(address):
    # Condition 1: Remove additional information after the address
    address = re.sub(r'\s*\([^)]*\)', '', address)

    # Condition 2: Replace variations of 'Ayer Rajah Food Centre' with 'Ayer Rajah Food Centre'
    if 'Ayer Rajah Food Centre' in address:
        address = 'Ayer Rajah Food Centre'

    # Condition 3: Replace variations of 'Ayer Rajah CC' with 'Ayer Rajah CC'
    elif 'Ayer Rajah Cc' in address or 'Ayer Rajah Community Centre' in address:
        address = 'Ayer Rajah CC'

    # Condition 4: Replace variations of 'Asia Square' with 'Asia Square'
    elif 'Asia Square' in address:
        address = 'Asia Square'

    # Replace variation 'Abc Brickworks Market & Food Centre Market & Food Centre Food Centre' 
    elif 'Abc Brickworks Market & Food Centre' in address:
        address = 'Abc Brickworks Market & Food Centre'

        
    # Condition 5: Replace 'Amk' with 'Ang Mo Kio'
    address = address.replace('Amk', 'Ang Mo Kio')

    #Replace variation 'Abc Brickworks Market & Food Centre Market & Food Centre Food Centre' 
    
    return address

# Apply the standardization function to the 'address' column of the DataFrame
establishment_df['address'] = establishment_df['address'].apply(standardize_address)


In [20]:
# Step 1: Filter the DataFrame
filtered_df = establishment_df[establishment_df['address'].str.lower().str.startswith(('a'))]

# Step 2: Extract distinct values from the filtered DataFrame
distinct_addresses_filtered = filtered_df['address'].unique()

# Step 3: Sort the distinct values alphabetically
distinct_addresses_filtered_sorted = sorted(distinct_addresses_filtered)

# Display the sorted distinct values
print("Distinct addresses starting with 'a':")
for address in distinct_addresses_filtered_sorted:
    print(address)

Distinct addresses starting with 'a':
A' Posh Bizhub
A'Posh Bizhub
Abc Brickworks Market & Food Centre
Acacia Breeze @ Yishun
Acacia Court
Academia
Academy Of Singapore Teachers
Ace Community Club
Adam Road Food Centre
Adda
Adelphi Park Estate
Admiralty
Admiralty Link
Admiralty Street
Admirax
Agape Village
Ah Hood
Ah Hood Road
Ah Wang Kway Chap
Air View Building
Airport Boulevard
Airview Building
Albert Centre
Alexandra
Alexandra Central
Alexandra Central Mall
Alexandra Hawker Centre
Alexandra Hospital
Alexandra Retail Centre
Alexandra Retail Centre Psa Building
Alexandra Road
Alexandra Technopark
Alexandra Village Food Centre
Aljunied
Aljunied 119 Food House
Aljunied Ave 2
Aljunied Ave 3
Aljunied Ave 4
Aljunied Ave 5
Aljunied Avenue 2
Aljunied B
Aljunied Crescent
Aljunied Industrial Complex
Aljunied Industrial Estate
Aljunied Mrt
Aljunied Mrt Station
Aljunied Road
Alkaff Crescent
Alkaff Mansion
Alkaff Vista
Alocassia
Alocassia Apartments
Alpha Industrial Building
Altez
American Grill


#### The distint values reduced further from original 4883 to 3971

In [21]:
#To find the total distinct value after standardizing the data with removal of '[xxx]' and change 'And' to '&'
total_distinct_addresses = establishment_df['address'].nunique()

# Display the total number of distinct addresses
print("Total distinct addresses:", total_distinct_addresses)

Total distinct addresses: 3971


### This is the standardization suggestion upon further investigation as its only done base on alphabet 'a', in future more cleaning needed to standardize the address to have better grouping

In [22]:
print(establishment_df.isnull().sum())

id_source                  0
name                       6
address                    0
country                    0
cuisine                    2
currency                  19
delivery_cost             51
lat                        0
lon                        0
opening_hours              0
image_url                 11
radius                     0
rating                  2132
reviews_nr              2534
delivery_options           0
promo                   7359
loc_type                   0
delivery_by                0
delivery_time            183
standardized_address       0
dtype: int64


#### Filter out corrupt Data . Removing rows with Null values in 'name' column 
- Possible reasons as the name shows test this might be test vendors at the initial grab phrase so dropping this rows

In [23]:
# Drop rows with null values in the 'name' column
establishment_df.dropna(subset=['name'],inplace=True)

In [24]:
print(establishment_df.isnull().sum())

id_source                  0
name                       0
address                    0
country                    0
cuisine                    2
currency                  19
delivery_cost             51
lat                        0
lon                        0
opening_hours              0
image_url                 11
radius                     0
rating                  2126
reviews_nr              2528
delivery_options           0
promo                   7354
loc_type                   0
delivery_by                0
delivery_time            183
standardized_address       0
dtype: int64


In [25]:
establishment_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16130 entries, 0 to 16135
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id_source             16130 non-null  object 
 1   name                  16130 non-null  object 
 2   address               16130 non-null  object 
 3   country               16130 non-null  object 
 4   cuisine               16128 non-null  object 
 5   currency              16111 non-null  object 
 6   delivery_cost         16079 non-null  float64
 7   lat                   16130 non-null  float64
 8   lon                   16130 non-null  float64
 9   opening_hours         16130 non-null  object 
 10  image_url             16119 non-null  object 
 11  radius                16130 non-null  int64  
 12  rating                14004 non-null  float64
 13  reviews_nr            13602 non-null  float64
 14  delivery_options      16130 non-null  object 
 15  promo                 87

In [26]:
# Filter the DataFrame to select rows with null values in the 'rating' column
establishment_df[establishment_df['rating'].isnull()].head(5)


id_source                  name         address    country  \
47  4-C2TYVNCKMAEBSA     Pet Lovers Centre        Vivocity  Singapore   
50  4-C2KWT4CGAK3DC2   Barossa Bar & Grill       Vivo City  Singapore   
64  4-C2UBV8MHHADJG6  Tree Side Nasi Lemak       Vivo City  Singapore   
73  4-C2TYTCLHWBVKHE        Ola Beach Club  Ola Beach Club  Singapore   
84  4-C2MTBB4FE6NVFE            Thai Ai Ai       Vivo City  Singapore   

                                              cuisine currency  delivery_cost  \
47  ["Mart", "Pet Accessories", "Pet Care", "Pet F...      SGD          570.0   
50                                        ["Western"]      SGD          570.0   
64                             ["Asian", "Malaysian"]      SGD          570.0   
73                           ["Beverages", "Western"]      SGD          760.0   
84                         ["Asian", "Local", "Thai"]      SGD          570.0   

         lat         lon                                      opening_hours  \
47  1.265074  103.822735  {"open": true, "displayedHours": "10:00-21:40"...   
50  1.265002  103.822709  {"open": true, "displayedHours": "11:30-21:00"...   
64  1.264305  103.822309  {"open": true, "displayedHours": "10:30-21:00"...   
73  1.251185  103.817939  {"open": true, "displayedHours": "10:00-19:30"...   
84  1.264305  103.822309  {"open": true, "displayedHours": "09:00-21:45"...   

                                            image_url  radius  rating  \
47  https://d1sag4ddilekf6.cloudfront.net/compress...    2500     NaN   
50  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     NaN   
64  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     NaN   
73  https://d1sag4ddilekf6.cloudfront.net/compress...    2500     NaN   
84  https://d1sag4ddilekf6.cloudfront.net/compress...    5000     NaN   

    reviews_nr   delivery_options                              promo loc_type  \
47         NaN      ONLY_DELIVERY  June Specials - Discounted Items!     MART   
50         NaN  DELIVERY_TAKEAWAY                                NaN     FOOD   
64         NaN      ONLY_DELIVERY                                NaN     FOOD   
73         NaN  DELIVERY_TAKEAWAY                                NaN     FOOD   
84         NaN  DELIVERY_TAKEAWAY    30% off with S$15.00 min. order     FOOD   

   delivery_by  delivery_time standardized_address  
47        GRAB           38.0             Vivocity  
50        GRAB           44.0            Vivo City  
64        GRAB           31.0            Vivo City  
73        GRAB           40.0       Ola Beach Club  
84        GRAB           39.0            Vivo City

In [27]:
# Step 2: Extract distinct values from the filtered DataFrame
establishment_df['loc_type'].unique()

array(['FOOD', 'MART'], dtype=object)

#### Drop columns to remove duplicate columns
- ignore the error as it was run twice

In [32]:
# List of columns to drop
columns_to_drop = ['country', 'cuisine', 'currency', 'delivery_cost', 'opening_hours', 'image_url', 'radius', 'delivery_options', 'promo', 'delivery_by', 'delivery_time']

# Drop the specified columns from the DataFrame
establishment_df.drop(columns=columns_to_drop, inplace=True)


In [ ]:
establishment_df.tail(5)

id_source                         name  \
16131  4-CZNFTYKKVA3UN2                 Minjiangkueh   
16132         SGDD01192  The Modern Izakaya by Tatsu   
16133         SGDD00961               Raj Restaurant   
16134  4-C2AHJVATN7C2RE   Albismi Indian Muslim Food   
16135  4-C2DXFEK2R6KJTJ  Zam Zam Indian Special Food   

                        address       lat         lon  rating  reviews_nr  \
16131          Ang Mo Kio Ave 3  1.370340  103.843225     5.0         2.0   
16132                       18)  1.295404  103.851698     NaN         NaN   
16133             Soho @ Farrer  1.310830  103.852873     3.9        72.0   
16134            Woodlands Road  1.396885  103.753440     3.6        19.0   
16135  Yishun Industrial Park A  1.441651  103.831330     4.3        37.0   

      loc_type      standardized_address  
16131     FOOD          Ang Mo Kio Ave 3  
16132     FOOD                       18)  
16133     FOOD             Soho @ Farrer  
16134     FOOD            Woodlands Road  
16135     FOOD  Yishun Industrial Park A

#### Noted there  was 2 columns for address: 'address' & 'standardized_address' 
- 'standardized_address' column was created as a result of trial and error in the process of data standardization
-  find the distict values of the both columns to chk if the standardization applied correctly to 'address' column

In [ ]:
#To find the total distinct value after standardizing the data with removal of '[xxx]' and change 'And' to '&'
total_distinct_addresses = establishment_df['standardized_address' ].nunique()

# Display the total number of distinct addresses
print("Total distinct addresses:", total_distinct_addresses)

Total distinct addresses: 4025


In [ ]:
#To find the total distinct value after standardizing the data with removal of '[xxx]' and change 'And' to '&'
total_distinct_addresses = establishment_df['address'].nunique()

# Display the total number of distinct addresses
print("Total distinct addresses:", total_distinct_addresses)

Total distinct addresses: 3967


#### Drop another column that was created as a result of trial and error in the process of data standardization

In [ ]:
# List of columns to drop
columns_to_drop = ['standardized_address']

# Drop the specified columns from the DataFrame
establishment_df.drop(columns=columns_to_drop, inplace=True)


In [ ]:
establishment_df.head(3)

id_source           name    address       lat         lon  rating  \
0         SGDD01085  Texas Chicken   Vivocity  1.264792  103.822103     3.8   
1  4-CYTDLPUJEP53N6         Subway   Vivocity  1.264741  103.822072     4.2   
2         SGDD05605       LiHO Tea  Vivo City  1.264309  103.821369     4.5   

   reviews_nr loc_type  
0      1591.0     FOOD  
1       197.0     FOOD  
2       353.0     FOOD

#### Change all entries under 'address' column to lowercase to standardized further for easier analytics process in the future

In [ ]:
# Convert the values in the 'address' column to lowercase
establishment_df['address'] = establishment_df['address'].str.lower()


In [ ]:
establishment_df.head(3)

id_source           name    address       lat         lon  rating  \
0         SGDD01085  Texas Chicken   vivocity  1.264792  103.822103     3.8   
1  4-CYTDLPUJEP53N6         Subway   vivocity  1.264741  103.822072     4.2   
2         SGDD05605       LiHO Tea  vivo city  1.264309  103.821369     4.5   

   reviews_nr loc_type  
0      1591.0     FOOD  
1       197.0     FOOD  
2       353.0     FOOD

### Download neccesay libraries to Load the  establishment dataframe to postgres database

In [ ]:
!pip install sqlalchemy
!pip install psycopg2

   ---------------------------------------- 0.0/1.2 MB ? eta -:--:--
   - -------------------------------------- 0.0/1.2 MB 1.4 MB/s eta 0:00:01
   -- ------------------------------------- 0.1/1.2 MB 825.8 kB/s eta 0:00:02
   ------------ --------------------------- 0.4/1.2 MB 3.2 MB/s eta 0:00:01
   ----------------------------- ---------- 0.9/1.2 MB 5.0 MB/s eta 0:00:01
   ---------------------------------------- 1.2/1.2 MB 5.7 MB/s eta 0:00:00


### Adding to a Database Grab_df : ESTABLISHMENT_TBL - SAILA

In [ ]:
from sqlalchemy import create_engine
username='postgres'
password = 'admin'
hostname = 'localhost'
port = 5432
database = 'grab_db'

db_url = f'postgresql://{username}:{password}@{hostname}:{port}/{database}'

#create a connection to the database
#engine = create_engine('postgresql://postgres:admin@localhost:5432/grab_db')
engine = create_engine(db_url)


# Create a new establishment table
commands = ('''Create TABLE IF NOT EXISTS establishment_tbl(id_source VARCHAR PRIMARY KEY,
                                                name VARCHAR,
                                                address VARCHAR,
                                                lat NUMERIC,
                                                lon NUMERIC,
                                                rating NUMERIC,
                                                reviews_nr NUMERIC,
                                                loc_type VARCHAR
                                                );''')

#Copy the data to the PostgreSQL table
establishment_df.to_sql('establishment_tbl', engine, if_exists='replace',index=False)

130

### Creating the dataframe : cuisine_df  

first we are going to replace the establishment_df['cuisine'] list rows from " into ' quote.
Then we create a column for the eventual primary key called cus_id
then we only have 3 columns in cuisine_df which is cuisine_id , id_source and cuisine

In [ ]:
cuisine_df=establishment_df

#clearing ' due to aprostophe 's from cuisine_df['cuisine'] list
cuisine_df['cuisine']=cuisine_df['cuisine'].str.replace('\'',"")

#clearing " from cuisine_df['cuisine'] list
cuisine_df['cuisine']=cuisine_df['cuisine'].str.replace('"',"")

#clearing [,] from cuisine_df['cuisine'] list
cuisine_df['cuisine']=cuisine_df['cuisine'].str.replace('[',"")
cuisine_df['cuisine']=cuisine_df['cuisine'].str.replace(']',"")

# Add a primary key column
cuisine_df['cuisine_id'] = ['cus_' + str(i + 1) for i in range(len(cuisine_df))]

cuisine_df=cuisine_df[['cuisine_id','id_source','cuisine']]

#making all text in cuisine column to lowercase
cuisine_df.loc[:, 'cuisine']=cuisine_df['cuisine'].str.lower()

cuisine_df

cuisine_id         id_source  \
0          cus_1         SGDD01085   
1          cus_2  4-CYTDLPUJEP53N6   
2          cus_3         SGDD05605   
3          cus_4         SGDD07548   
4          cus_5  4-CY2ZLZLTT2EDJ2   
...          ...               ...   
16131  cus_16126  4-CZNFTYKKVA3UN2   
16132  cus_16127         SGDD01192   
16133  cus_16128         SGDD00961   
16134  cus_16129  4-C2AHJVATN7C2RE   
16135  cus_16130  4-C2DXFEK2R6KJTJ   

                                                 cuisine  
0          western, fast food, halal, chicken, beverages  
1         healthy, sandwiches, fast food, halal, western  
2      same prices in-store, beverages, local, bubble...  
3             halal, chicken, fast food, burger, western  
4                        japanese, ramen, noodles, asian  
...                                                  ...  
16131  asian, bakery & cake, breakfast & brunch, sing...  
16132                           japanese, noodles, grill  
16133                        indian, dessert, vegetarian  
16134                                             indian  
16135                                             indian  

[16130 rows x 3 columns]

null values found in cuisine, this we will fill it with thai for index 2473 and korean, beveragees for indx 15049

In [ ]:
cuisine_df[cuisine_df['cuisine'].isna()]

cuisine_id         id_source cuisine
2473    cus_2468  4-CZB3EYJJMEUBAX     NaN
15049  cus_15044  4-C2EVG4BDGEVTG6     NaN

filling up null values in cuisine column with appropriate category types according to their establishment information

In [ ]:
# index 2473 filling up null values with relevant values
cuisine_df.at[2473, 'cuisine'] = 'thai'

# index 15049 filling up null values with relevant values
cuisine_df.at[15049, 'cuisine'] = 'beverages, korean'

#you may check again if there are any more nulls in cuisine column
#cuisine_df[cuisine_df['cuisine'].isna()]


cuisine_id         id_source            cuisine
15055  cus_15050  4-CY3XNTJKENNXNE  indian, malaysian

For each row sort the values in the cuisines category in alphabetical order a-z

In [ ]:
# Sort the values within each string alphabetically
cuisine_df['cuisine'] = cuisine_df['cuisine'].apply(lambda x: ', '.join(sorted(x.split(', '))))


cuisine_df

C:\Users\T470\AppData\Local\Temp\ipykernel_23004\3522236522.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cuisine_df['cuisine'] = cuisine_df['cuisine'].apply(lambda x: ', '.join(sorted(x.split(', '))))


cuisine_id         id_source  \
0          cus_1         SGDD01085   
1          cus_2  4-CYTDLPUJEP53N6   
2          cus_3         SGDD05605   
3          cus_4         SGDD07548   
4          cus_5  4-CY2ZLZLTT2EDJ2   
...          ...               ...   
16131  cus_16126  4-CZNFTYKKVA3UN2   
16132  cus_16127         SGDD01192   
16133  cus_16128         SGDD00961   
16134  cus_16129  4-C2AHJVATN7C2RE   
16135  cus_16130  4-C2DXFEK2R6KJTJ   

                                                 cuisine  
0          beverages, chicken, fast food, halal, western  
1         fast food, halal, healthy, sandwiches, western  
2      beverages, bubble tea, local, same prices in-s...  
3             burger, chicken, fast food, halal, western  
4                        asian, japanese, noodles, ramen  
...                                                  ...  
16131  asian, bakery & cake, breakfast & brunch, sing...  
16132                           grill, japanese, noodles  
16133                        dessert, indian, vegetarian  
16134                                             indian  
16135                                             indian  

[16130 rows x 3 columns]

After sort alphabetically break into seperate columns

In [ ]:
# Split the 'cuisine' column by comma and expand into separate columns
cuisines_split = cuisine_df['cuisine'].str.split(', ', expand=True)

# Rename the new columns
cuisines_split.columns = [f'cuisine_{i+1}' for i in range(cuisines_split.shape[1])]

# Concatenate the new columns with the original DataFrame and drop the old cuisine column
cuisine_df = pd.concat([cuisine_df.drop(columns=['cuisine']), cuisines_split], axis=1)

cuisine_df

cuisine_id         id_source  cuisine_1      cuisine_2  \
0          cus_1         SGDD01085  beverages        chicken   
1          cus_2  4-CYTDLPUJEP53N6  fast food          halal   
2          cus_3         SGDD05605  beverages     bubble tea   
3          cus_4         SGDD07548     burger        chicken   
4          cus_5  4-CY2ZLZLTT2EDJ2      asian       japanese   
...          ...               ...        ...            ...   
16131  cus_16126  4-CZNFTYKKVA3UN2      asian  bakery & cake   
16132  cus_16127         SGDD01192      grill       japanese   
16133  cus_16128         SGDD00961    dessert         indian   
16134  cus_16129  4-C2AHJVATN7C2RE     indian           None   
16135  cus_16130  4-C2DXFEK2R6KJTJ     indian           None   

                cuisine_3             cuisine_4 cuisine_5  
0               fast food                 halal   western  
1                 healthy            sandwiches   western  
2                   local  same prices in-store      None  
3               fast food                 halal   western  
4                 noodles                 ramen      None  
...                   ...                   ...       ...  
16131  breakfast & brunch           singaporean      None  
16132             noodles                  None      None  
16133          vegetarian                  None      None  
16134                None                  None      None  
16135                None                  None      None  

[16130 rows x 7 columns]

### Checking commonality in the grouping to see if we can clean further

In [ ]:
#check commonlity

# Specify selected columns
selected_columns = ['cuisine_1', 'cuisine_2', 'cuisine_3', 'cuisine_4', 'cuisine_5']
# Apply value_counts function to each column and calculate commonality
commonality = cuisine_df[selected_columns].apply(pd.Series.value_counts, axis=1).fillna(0).sum().astype(int)

print(commonality)

1 michelin star         11
2 michelin stars         1
accessories              2
alcohol                464
alcoholic beverages      8
                      ... 
wine                    67
wings                    7
wok                      7
wraps                   12
yogurt                  26
Length: 291, dtype: int32


Further cleaning to group certain category like michelin star, alcohol and etc

In [ ]:
# further cleaning to change 2 michelin star to 1 michelin star

# Find rows where any of the selected columns contain '2 michelin stars'
rows_to_replace = cuisine_df[selected_columns].apply(lambda x: x.str.contains('2 michelin stars')).any(axis=1)
# Replace '2 michelin stars' with 'michelin star'
cuisine_df.loc[rows_to_replace, selected_columns] = cuisine_df[selected_columns].replace('2 michelin stars', '1 michelin star')

# further cleaning to change 1 michelin star to michelin star
rows_to_replace = cuisine_df[selected_columns].apply(lambda x: x.str.contains('1 michelin star')).any(axis=1)
# Replace '2 michelin stars' with 'michelin star'
cuisine_df.loc[rows_to_replace, selected_columns] = cuisine_df[selected_columns].replace('1 michelin star', 'michelin star')

# further cleaning to change michelin guide to michelin star
rows_to_replace = cuisine_df[selected_columns].apply(lambda x: x.str.contains('michelin guide')).any(axis=1)
# Replace '2 michelin stars' with 'michelin star'
cuisine_df.loc[rows_to_replace, selected_columns] = cuisine_df[selected_columns].replace('michelin guide', 'michelin star')

#further cleanin alcoholic beverages to just alcohol
rows_to_replace = cuisine_df[selected_columns].apply(lambda x: x.str.contains('alcoholic beverages')).any(axis=1)
# Replace 'alcoholic beverages' with 'alcohol'
cuisine_df.loc[rows_to_replace, selected_columns] = cuisine_df[selected_columns].replace('alcoholic beverages', 'alcohol')

#to group and count each unique value of cuisine column
# Apply value_counts function to each column and calculate commonality
commonality = cuisine_df[selected_columns].apply(pd.Series.value_counts, axis=1).fillna(0).sum().astype(int)

# Sort commonality counts from biggest to smallest
commonality_sort_values = commonality.sort_values(ascending=False)

# Sort commonality counts alphabetically
commonality_sort_alpha = commonality.sort_index()

### Adding to a Database Grab_df : CUISINE_TBL - HASSAN

In [ ]:
!pip install sqlalchemy
!pip install psycopg2

In [ ]:
from sqlalchemy import create_engine
username='postgres'
password = 'admin'
hostname = 'localhost'
port = 5432
database = 'grab_db'

db_url = f'postgresql://{username}:{password}@{hostname}:{port}/{database}'

#create a connection to the database
#engine = create_engine('postgresql://postgres:admin@localhost:5432/grab_db')
engine = create_engine(db_url)


# Create a new cuisine table
commands = ('''Create TABLE IF NOT EXISTS cuisine_tbl(cuisine_id VARCHAR PRIMARY KEY,
                                                id_source VARCHAR,
                                                streaming_service VARCHAR,
                                                cuisine_1 VARCHAR,
                                                cuisine_2 VARCHAR,
                                                cuisine_3 VARCHAR,
                                                cuisine_4 VARCHAR,
                                                cuisine_5 VARCHAR,
                                                );''')

# Step 9: Copy the data to the PostgreSQL table
cuisine_df.to_sql('cuisine_tbl', engine, if_exists='replace', index=False)

130

### establishment opening hours dataframe - LEE HONG ENG